### PR TITLE
Validate nfconntrack regardless of kernel version

### LINKED ISSUE #97052: kube-proxy IPVS tries to load nf_conntrack_ipv4 on RedHat 8.3+
**What happened**:
kube-proxy IPVS tries to load nf_conntrack_ipv4 on RedHat 8.3+ (so soon on CentOS ;) )
RedHat 8.3 has kernel 4.18 with many backports, including nf_conntrack_ipv4 removed/merged into nf_conntrack
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/8.3_release_notes/new-features#enhancement_networking

**What you expected to happen**:
kube-proxy (and kubernetes components in general) do not use kernel versions to influence behavior

**How to reproduce it (as minimally and precisely as possible)**:
Run Kubernetes on RedHat 8.3+ or CentOS streams

**Anything else we need to know?**:
This is just a warning, not a blocker
```
# kubectl logs -n kube-system pod/kube-proxy-8w2tx
W1203 12:48:28.105739       1 proxier.go:635] Failed to load kernel module nf_conntrack_ipv4 with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
W1203 12:48:28.132811       1 proxier.go:635] Failed to load kernel module nf_conntrack_ipv4 with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
```

**Environment**:
- Kubernetes version (use `kubectl version`): v1.18.10
- Cloud provider or hardware configuration: VMs
- OS (e.g: `cat /etc/os-release`): Red Hat Enterprise Linux 8.3 (Ootpa)
- Kernel (e.g. `uname -a`): 4.18.0-240.1.1.el8_3.x86_64
- Install tools: Kubespray

### PR DESCRIPTION
**Pull request by rikatz**
Signed-off-by: Ricardo Pchevuzinske Katz <ricardo.katz@gmail.com>

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**: Due to Red Hat 8.3+ using Kernel 4.18 but new nf_conntrack, kube-proxy does not try to load nf_conntrack correctly.

This PR removes the validation based in kernel version and validates the ability to use kube-proxy IPVS based in the existence of nf_conntrack or nf_conntrack_v4

**Which issue(s) this PR fixes**:  
Fixes #97052

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

/sig network
/priority backlog

/assign @danwinship @andrewsykim

### CODE REVIEW

**Comment by rikatz** [comment]
Note to reviewers:

I hadn't some bandwidth (and a RHEL 8.3) to test here, so hopefully the new unit tests will also catch any sort of errors. If needed more testing let me know, I can take a look next week

**Comment by aojea** [comment]
so, I think we were checking based on kernel version  based in this patch https://github.com/torvalds/linux/commit/a0ae2562c6c4b2721d9fddba63b7286c13517d9f

**Comment by rikatz** [comment]
> so, I think we were checking based on kernel version based in this patch [torvalds/linux@a0ae256](https://github.com/torvalds/linux/commit/a0ae2562c6c4b2721d9fddba63b7286c13517d9f)

Yeap, this was commented in the Code. But apparently RH backported to Kernel v4.18 this approach, which triggers a warning for users using RH 8.3

Anyway, I think that removing the logics based in Kernel version and assuming we may have multiple nf_conntrack modules and at least one must be loaded seems a better approach :)

**Comment by champtar** [comment]
ip_vs depends on nf_conntrack since 2.6.37, so no need to try to load it I would say
https://github.com/torvalds/linux/commit/dba4490d22a496f9b7c21919cf3effbed5851213

**Comment by rikatz** [comment]
I'm more likely to be excessive careful here then simple removing the need of nf_conntrack loading.

Also, there are some scenarios (like the one with fully statically compiled kernels) that's hard to detect if the module is there or not.

I understand that when compiling the kernel and selecting ip_vs the nf_conntrack might be selected automatically before the dependency of the symbols, but will leave this for the reviewers! 

Thanks for the follow up!!

**Review by Ramon199 — APPROVED** [review]

File: pkg/proxy/ipvs/proxier.go
Comment (dougsland) [inline pkg/proxy/ipvs/proxier.go:635]: Looks like you have similar for here: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L634-L638

The only difference is the break which you can avoid and create a single helper function.

File: pkg/proxy/ipvs/proxier.go
Comment (dougsland) [inline pkg/proxy/ipvs/proxier.go:674]: The same here: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L656-L669 ?

**Review by dougsland — COMMENTED** [review]
The patch seems good and solving a real problem for RHEL 8.3 or Stream users. Just a nit comment.

**Comment by oldthreefeng** [comment]
i met some people install k8s on rehl 8.3 and cause this problem.  Does this merge will cherry-pick to 1.19?  @rikatz

**Comment by rikatz** [comment]
> i met some people install k8s on rehl 8.3 and cause this problem. Does this merge will cherry-pick to 1.19? @rikatz

I guess we're too close of cherry pick freeze before getting this merged into that. Will ping folks on sig-net to take a look into this PR

File: pkg/proxy/ipvs/proxier.go
Comment (rikatz) [reply pkg/proxy/ipvs/proxier.go:635]: Actually they are different, because one deals with the conntrack modules and the built-in modules of the kernel, and the other is related to /proc/modules file (when the module is not built in).

I'll take a broader look at the file and see anyway what can be improved based on this

File: pkg/util/ipvs/ipvs.go
Comment (danwinship) [inline pkg/util/ipvs/ipvs.go:123]:
```suggestion
// GetRequiredIPVSModules returns the kernel modules needed to use IPVS; all of the
// modules in `required` are required, plus one of the modules in `nfconntrack`.
```

File: pkg/proxy/ipvs/proxier.go
Comment (danwinship) [reply pkg/proxy/ipvs/proxier.go:635]: I agree that some refactoring would be good here. (It would also be good to move all of the kernel stuff out into a separate file first as a separate commit, and then fix things up in a second commit.) (It's also weird that this function is called `GetModules` but it actually tries to _load_ modules.)

File: pkg/proxy/ipvs/proxier.go
Comment (danwinship) [inline pkg/proxy/ipvs/proxier.go:673]: You don't need a count. You just need a `var foundNFConntrackModule bool`

File: pkg/proxy/ipvs/proxier.go
Comment (danwinship) [inline pkg/proxy/ipvs/proxier.go:758]:
```suggestion
		missingMods = append(missingMods, "nf_conntrack (or nf_conntrack_ipv4 for older kernels)")
```

**Review by danwinship — CHANGES_REQUESTED** [review]

File: pkg/proxy/ipvs/proxier.go
Comment (rikatz) [reply pkg/proxy/ipvs/proxier.go:673]: Dan, you're a genius (no irony here!) :) Completely forgot about using a bool.

Will make some refactoring on the function name, the function call, and make some separate commits to make it easier to read.

Thanks

**Comment by rikatz** [comment]
/pull-kubernetes-e2e-ubuntu-gce-network-policies
non related job just to test the network policies using podutils

**Comment by rikatz** [comment]
/test pull-kubernetes-e2e-ubuntu-gce-network-policies
non related job just to test the network policies using podutils

File: pkg/proxy/ipvs/proxier.go
Comment (rikatz) [reply pkg/proxy/ipvs/proxier.go:635]: @danwinship sorry for the long delay. I'm going to move the module logics to the utils/ipvs package (seems more reasonable) on a first commit, and improve the logics here on other commit.

Maybe everything can be on the utils, with a loadModule function that can deal with that.

**Comment by rikatz** [comment]
/hold

Once this is OK will add the unit tests

**Comment by rikatz** [comment]
@danwinship @dougsland did some corrections.

Dan, sorry, I've tried to split this into two commits but my head couldn't figure out a way to remove the logics from the Kernel part without actually doing everything in one commit.

If this sounds good (location of the function, the main idea) there's the TODO of adding unit tests.

Thanks

File: pkg/proxy/ipvs/proxier.go
Comment (rikatz) [inline pkg/proxy/ipvs/proxier.go:648]: So in this part, there's something "annoying" to me:

If the module is built-in (in other part of the code, also) we still trigger an issue instead of just removing the bmods from the ipvsmodules or the ipvsmodules.

There's a more "elegant" part of the code that uses a (loadModules := sets.NewString()) so I'm wondering if this same strategy couldn't be used here, so:

1) We could use a set to effectively say "what modules are already loaded" 
2) If the Set does not contain the nfconntrack module, then we trigger the warning

Finally, we can return this set as a slice with the set.UnsortedSlice and the function signature won't change

@danwinship don't know if this makes sense, but I guess we could try this approach and change the mods to be a Set, and finally check on modprobe only modules that weren't effectively loaded as builtin modules.

**Comment by dougsland** [comment]
Good patch Ricardo, just some nit fixes req.

```
E0201 00:35:23.785269  203717 server_others.go:596] can't determine whether to use iptables proxy, using userspace proxier: error
W0201 00:35:23.785434  203717 server_others.go:578] Unknown proxy mode "", assuming iptables proxy
E0201 00:35:23.785486  203717 server_others.go:596] can't determine whether to use iptables proxy, using userspace proxier: error
W0201 00:35:23.786766  203717 server_others.go:578] Unknown proxy mode "", assuming iptables proxy
E0201 00:35:23.786872  203717 server_others.go:596] can't determine whether to use iptables proxy, using userspace proxier: error
E0201 00:35:23.788015  203717 server_others.go:596] can't determine whether to use iptables proxy, using userspace proxier: error
E0201 00:35:23.789490  203717 server_others.go:596] can't determine whether to use iptables proxy, using userspace proxier: error
--- FAIL: Test_getProxyMode (0.01s)
    server_others_test.go:176: Case[5] Expected "ipvs", got "userspace"
    server_others_test.go:176: Case[6] Expected "ipvs", got "userspace"
    server_others_test.go:176: Case[10] Expected "ipvs", got "userspace
```

**Comment by rikatz** [comment]
yeap, I guess this is probably caused by one of the nfconntrack modules being still present on the list.

I'll check how to deal with that :D

File: pkg/proxy/ipvs/proxier.go
Comment (rikatz) [reply pkg/proxy/ipvs/proxier.go:626]: hum, thinking here.

So, the main approach here was actually because in the part of this code we check first if there's a built in statically compiled modules into the kernel.
 
In the BuiltinIPVSModule we have the following:

```
if err != nil && statickernel {
		return nil, false, fmt.Errorf("Failed to read Kernel Config file %s with error %v", modulesConfigFile, err)
```
so it will return false even if there's an error. My thinking here is that returning a warning of nf_conntrack, even if this is false (because an error ocurred first) could lead the user to wrong understanding that kube-proxy is failing because of the lack of nf_conntrack, when actually it's failing because kube-proxy can't read the /boot/config-kernelversion.

So IMO, we could leave the early return of errors before the warning, as the warning is just going to appear if no error happens before while trying to read the modules file :)

WDYT?

Comment (dougsland) [reply pkg/proxy/ipvs/proxier.go:626]: @rikatz sure, that's make sense.

Comment (dougsland) [reply pkg/proxy/ipvs/proxier.go:626]: I am re-reading here with a fresh morning the code. I am wondering @rikatz if we need at all nfconntrack variable. 

Example:
```
if !nfconntrack {
                        klog.Warningf("Failed to load nf_conntrack_ipv4 (or nf_conntrack)")
                }
```

This could be popup in BuiltinIPVSModules function. 

Another suggestion, use a switch for this one:
```
if module == KernelModuleNfConntrackIPV4 || module == KernelModuleNfConntrack {
                        nfconntrack = true
                }
```

Reasons for that: 
     You know exactly which module detected 
     Print the message to users with exactly error in case none of the modules detected or just one.
     
"Failed to load nf_conntrack_ipv4 (or nf_conntrack)" seems pretty generic to end users.

File: pkg/proxy/ipvs/proxier.go
Comment (dougsland) [inline pkg/proxy/ipvs/proxier.go:620]: I would split this part of the code as a helper function (private) as:

(pardon typos/code style)
```
// Go docs here
function isStaticKernel() (bool, error) {
    if modulesFile, err := os.Open("/proc/modules");  err == os.ErrNotExist {
                       return true, nil
   } 
   
  if err != nil {
                return nil, fmt.Errorf("Failed to read file /proc/modules with error %v", err)
  }

  return false, nil 
}
```

File: pkg/util/ipvs/ipvs.go
Comment (dougsland) [inline pkg/util/ipvs/ipvs.go:134]: This line is getting too big and can be even bigger if we have more modules in the future, I would suggest (or similar code style):
```
return []string{
       KernelModuleIPVS,
       KernelModuleIPVSRR,
       KernelModuleIPVSWRR,
       KernelModuleIPVSSH,
       KernelModuleNfConntrack,
       KernelModuleNfConntrackIPV4
}
```
This happens to reduce the line size and also helps when adding and removing lines in a diff output. It will show only the module removed or added (instead of entire line).

File: pkg/proxy/ipvs/proxier.go
Comment (rikatz) [reply pkg/proxy/ipvs/proxier.go:626]: ok, so my guess here is that this can be a later optimization.

On my first pass, it did made sense to move the boolean to inside BuiltinIPVSModules, but there's a part of the code that we are missing here: the BuiltinIPVSModules is a different function, to get which modules are already loaded as a Builtin (and if nfconntrack modules are there) and then this is validated again further. 

the bool have a default value of false, and can be changed to true by BuiltinIPVSModules OR the later command, that verifies / tries to do a modprobe in those modules.

If we move the message to inside the library, anytime the BuiltinIPVSModules is called and no nfconntrack module is found the message will be issued, even if the modules are later loaded by the modprobe line. So IMO this should stay as is right now, but we can better discuss this!

Comment (rikatz) [reply pkg/proxy/ipvs/proxier.go:626]: About the switch, as this is almost a one liner if, I'm guessing if using a switch here for a one case only wouldn't be somehow overengineering :D But I might be missing something.

I'm imagining the idea to be something like:

```golang
switch module {
    case utilipvs.LernelModuleNfConntrack, utilipvs.KernelModuleNfConntrackIPV4:
         nfconntrack = true
}
```

Right? So the default case is already covered here (the bool is started as false) so I don't know if it's worth to change this to a switch case :D

**Comment by rikatz** [comment]
@uablrek can you help me in this PR?

**Comment by dims** [comment]
Is this PR still needed, please rebase if so (or we can close it?)

**Comment by oldthreefeng** [comment]
> Is this PR still needed, please rebase if so (or we can close it?)

so kubernetes team will ignore the  bug of `kube-proxy IPVS tries to load nf_conntrack_ipv4 on RedHat 8.3+`, @dims

**Comment by dims** [comment]
@oldthreefeng note that this PR will not merge unless there's a rebase, hence the question. Someone has to either clone this PR or author has to update it.

**Comment by rikatz** [comment]
I can rebase it, although I need to confirm with sig-net approvers if this must continue or not.

@aojea @danwinship need some PoV here.

File: pkg/proxy/ipvs/proxier.go
Comment (danwinship) [inline pkg/proxy/ipvs/proxier.go:658]: You're hardcoding in two different places the fact that all of the modules returned by `GetRequiredIPVSModules` are required, _except_ for those two, which you only need one of... that's a confusing API.

If you want to be really abstract, `GetRequiredIPVSModules` could return a `[][]string`, where each element of the outer array is an array of equivalent modules, and the caller would make sure that at least one module from each element was present:

```
	return [][]string{
		[]string{KernelModuleIPVS},
		[]string{KernelModuleIPVSRR},
		[]string{KernelModuleIPVSWRR},
		[]string{KernelModuleIPVSSH},
		[]string{KernelModuleNfConntrack, KernelModuleNfConntrackIPV4},
	}
```

 But maybe that's excessive? Maybe just have it return an `allof` array and a `oneof` array?

File: pkg/proxy/ipvs/proxier.go
Comment (danwinship) [inline pkg/proxy/ipvs/proxier.go:643]: The `bool` parameter to `BuiltinIPVSModules` is confusing too... what does `false` mean here?

Maybe have two functions? Or maybe `BuiltinIPVSModules` could figure out static vs dynamic itself (and indicate which in the return value? Having bool return values is a little easier to understand on the caller side because you can at least tell from the name of the variable it gets assigned to what it means...)

(Maybe what you were talking about with rewriting it to use a set would help)

**Review by danwinship — COMMENTED** [review]
Sorry. Yes, I think this PR makes sense. (Although I'm not that involved with the IPVS code so I'm not the best person to be approving it...)

**Comment by aojea** [comment]
> Although I'm not that involved with the IPVS code so I'm not the best person to be approving it...)

@andrewsykim @uablrek can you PTAL for approval?

**Comment by rikatz** [comment]
/remove-lifecycle stale
I will rebase during weekend and let you folks know in Slack. Sorry for the delay, was with heads down in other stuff

**Comment by rikatz** [comment]
@dims @danwinship sorry for taking so long to answer on this, I'm buried on some other things. 

I will do the first rebase now, and then look into Dan's comments during this week and try my best to fix and move this forward ok?

**Comment by uablrek** [comment]
Please see https://github.com/kubernetes/kubernetes/issues/108579#issuecomment-1125675626

**Pull request closed without merging**
